WalletWasabi.Tests/UnitTests/Clients/WhaleCoinjoinTests.cs

```diff
@@ -0,0 +1,229 @@
+using NBitcoin;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using WalletWasabi.Blockchain.Analysis;
+using WalletWasabi.Blockchain.Keys;
+using WalletWasabi.Tests.Helpers;
+using WalletWasabi.WabiSabi.Client;
+using WalletWasabi.WabiSabi.Models;
+using WalletWasabi.WabiSabi;
+using Xunit;
+using WalletWasabi.Helpers;
+using WalletWasabi.Extensions;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.WabiSabi.Backend.Rounds;
+using DynamicData;
+using WalletWasabi.Blockchain.TransactionOutputs;
+using Xunit.Abstractions;
+using System.IO;
+using System.Threading;
+
+namespace WalletWasabi.Tests.UnitTests.Clients;
+
+public class WhaleCoinjoinTests
+{
+	private ITestOutputHelper TestOutputHelper { get; }
+	private string OutputFilePath { get; }
+
+	public WhaleCoinjoinTests(ITestOutputHelper testOutputHelper)
+	{
+		TestOutputHelper = testOutputHelper;
+		var outputFolder = Directory.CreateDirectory(Common.GetWorkDir(nameof(WhaleCoinjoinTests), "Output"));
+		var date = DateTime.Now.ToString("Md_HHmmss");
+		OutputFilePath = Path.Combine(outputFolder.FullName, $"Outputs{date}.txt");
+	}
+
+	[Theory]
+	[InlineData(1, 0.0005, 0.05, 1234)]
+	[InlineData(1, 0.0005, 0.05, 6789)]
+	[InlineData(1, 0.001, 0.1, 1234)]
+	[InlineData(1, 0.0005, 0.2, 1234)]
+	[InlineData(1, 0.0005, 0.01, 1234)]
+	public void ScoreForWhales(double whaleAmountBtc, double otherClientsAmountMin, double otherClientsAmountMax, int randomSeed)
+	{
+		var nbOtherClients = 30;
+		var otherNbInputsPerClient = 5;
+		var mockSecureRandom = new TestRandomSeed(randomSeed);
+
+		var anonScoreTarget = 100;
+		var maxTestRounds = 1000;
+		var displayWhaleCoinsEachRounds = int.MaxValue;
+
+		var analyser = new BlockchainAnalyzer();
+		var whaleMinInputAnonSet = 1.0;
+
+		RoundParameters roundParams = CreateMultipartyTransactionParameters();
+		KeyManager km = KeyManager.CreateNew(out _, "", Network.RegTest);
+		HdPubKey hdPub = BitcoinFactory.CreateHdPubKey(km);
+		HdPubKey[] otherHdPub = new HdPubKey[nbOtherClients];
+		for (int i = 0; i < nbOtherClients; i++)
+		{
+			KeyManager kmTmp = KeyManager.CreateNew(out _, "", Network.RegTest);
+			otherHdPub[i] = BitcoinFactory.CreateHdPubKey(kmTmp);
+		}
+
+		var whaleSmartCoins = new List<SmartCoin> { BitcoinFactory.CreateSmartCoin(hdPub, (decimal)whaleAmountBtc) };
+		var totalVSizeWhale = 0;
+		var counter = 0;
+		while (whaleMinInputAnonSet < anonScoreTarget && counter <= maxTestRounds)
+		{
+			counter++;
+			var whaleSelectedSmartCoins = SelectCoinsForRound(whaleSmartCoins, anonScoreTarget, roundParams, mockSecureRandom);
+			var whaleSelectedCoins = whaleSelectedSmartCoins.Select(sm => sm.Coin);
```

Comment:
I'm not sure why you have two different variable for this.
`whaleSelectedCoins` feels redundant, since I can just replace every occuring of it with `whaleSelectedSmartCoins.Select(sm => sm.Coin);` and it will work IMO.